docs/proposals/hpa-v2.md

```diff
@@ -0,0 +1,242 @@
+apiVersion: autoscaling/v2alpha1
+spec:
+  scaleTargetRef:
+    kind: ReplicationController
+    name: WebFrontend
+  minReplicas: 2
+  maxReplicas: 10
+  targetCPUUtilizationPercentage: 80
+  metrics:
+  - object:
+      target:
+        kind: Service
+        name: Frontend
+      metricName: hits-per-second
+      targetValue: 1k
+```
+
+### Alternatives and Future Considerations ###
+
+Since the new design mirrors volume plugins (and similar APIs), it makes
+it relatively easy to introduce new fields in a backwards-compatible way:
+we simply introduce a new field in `MetricSpec` as a new "metric type".
+
+#### External ####
+
+As a potential future consideration, we may wish to include a metric
+source type of `External`, which simply has an opaque metric type and
+target value (as strings).  This would allow custom autoscalers with
+custom source mechanics to be written, while still providing the same
+indication that a controller was being autoscaled.
+
+#### Resource Percentages ####
+
+In the current design, CPU percentage metrics are a completely separate
+field from the rest of the metrics.  While this simplifies the common case
+of scaling on CPU, it makes the overall design a bit more clunky.  It also
+doesn't provide a method for scaling based on memory percentage.
+
+Two alternatives to the method proposed above are shown the object below.
+The former is more compact, but precludes scaling memory as a percentage
+of request.
+
+```yaml
+kind: HorizontalPodAutoscaler
+apiVersion: autoscaling/v2alpha1
+spec:
+  scaleTargetRef:
+    kind: ReplicationController
+    name: WebFrontend
+  minReplicas: 2
+  maxReplicas: 10
+  metrics:
+  - cpuRequest:
+      targetPercentage: 80
+  - resourceRequest:
+      resource: cpu
+      targetPercentage: 80
+```
+
+#### Referring to the current Namespace ####
+
+It is beneficial to be able to refer to a metric on the current namespace,
+similarly to the `ObjectMetricSource` source type, but without an explicit
+name.  Because of the similarity to `ObjectMetricSource`, it may simply be
+sufficient to allow specificying a `kind` of "Namespace" without a name.
+Alternatively, a similar source type to `PodsMetricSource` could be used.
+
+#### Aggregating Desired Replica Counts ####
+
+In this iteration of the proposal, we simply take the maximum of all the
+desired replica counts across the specified metrics to figure out the
+final replica count for the HPA.  However, in certain cases, it may be
+advantageous to instead take the minimum or average.  It could be
+desirable to add a field for specifying this (with "max" then becoming the
+default).
+
+Mechanical Concerns
+-------------------
+
+The HPA will derive metrics from two sources: resource metrics (i.e. CPU
+request percentage) will come from the
+[master metrics API](resource-metrics-api.md), while other metrics will
+come from the [custom/arbitrary metrics API](custom-metrics-api.md)
+(currently proposed as #34586).
```

Comment:
... which will be a private adapter to the monitoring solution being currently used in the cluster.